### DISCUSSION TITLE
right/wrong way to read from file in parallel?

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#23459] [Q&A General] [answered]
I have a large data file that feeds my AuxKernel (temperature field data, for example). It's a large model and I want to run in parallel. Is there a way I can read the file once and then send all that data to all my CPU's? Right now I've got the file-reading code in `initialSetup` of my AuxKernel, and it appears to be happening on all CPU's when I run in parallel, which doesn't seem very efficient and creates a bottleneck.

**Comment by GiudGiud** [comment]
Hello

This is possible.
In your intialSetup(), check for processor_id() == 0 before reading. 

Then once you have read on rank 0, use the MPI communicator to broadcast. It's usually a class attribute named _comm. 

See the doxygen for that class
https://mooseframework.inl.gov/docs/doxygen/timpi/classlibMesh_1_1Parallel_1_1Communicator.html
You ll like want something like "broadcast" for a rank 0 to all others communication. 

Guillaume

**Reply by GiudGiud** [reply]
note that depending on your system, distributing the file and reading only the part that matters to each rank can be a faster solution. It's likely that a hybrid solution, with file reading in some rank then communication to groups of other ranks is the fastest.

**Reply by jessecarterMOOSE** [reply]
Thanks, I'll look into that. What kinds of data are supported? Can I just distribute basic types like floats and strings or can I make a std::map and broadcast that?

**Reply by GiudGiud** [reply]
I see a map_broadcast so you can try maps.
I m not sure where the full list of instanciations of these templated communication routines are. 
@lindsayad ?

**Reply by jessecarterMOOSE** [reply]
Are there any (simple) examples of this? If process 0 broadcasts some data, how do other processes receive it?

**Reply by GiudGiud** [reply]
There's a broadcast of some tuple in MultiAppTransfer.C to communicate the multiapp coord transform

**Reply by jessecarterMOOSE** [reply]
looks like map_broadcast doesn't take a std::map but instead a Map type. Any idea how one converts to/from that? Also the function takes a few other arguments which I can't figure out.

**Reply by jessecarterMOOSE** [reply]
Took a guess at the other arguments (0 and true) and the compiler complains that the map_broadcast function is private. I do see a few overloaded versions of the function in the doxygen, but I can't tell the difference. They all take a Map, unsigned int, and bool.

**Reply by GiudGiud** [reply]
you call broadcast and under the hood it links to map_broadcast.
The broadcast call should be very simple

_comm.broadcast(the thing)

**Reply by jessecarterMOOSE** [reply]
oh, thanks for pointing that out

**Reply by jessecarterMOOSE** [reply]
ok, new error. The compiler doesn't seem to like my map, which is `std::map<std::vector<Real>, const Function *>`. (changing the map's return type to a Real does compile though, so we're getting close)

**Reply by jessecarterMOOSE** [reply]
```
timpi/packing.h:256:38: error: 'packable_size' is not a member of 'libMesh::Parallel::Packing<const Function*, void>'

256 |       return Packing<T3>::packable_size<comp, ctx);
```

I'm guessing it can't determine the size of the data I'm trying to send?

**Reply by jessecarterMOOSE** [reply]
but what I can do is store the _name_ (std::string) of the Function in the map instead of a pointer to the function, and then in computeValue() use `getFunctionByName()`. Do you see any problems with that? Is that particularly slow?

**Reply by GiudGiud** [reply]
Function or Function* is not allowed
How would we transfer that
A pointer would be invalid on the target? A function is a complex object?

**Reply by GiudGiud** [reply]
the name seems fine to me

**Reply by lindsayad** [reply]
It seems like you're on the right track here.

I have to wonder though whether communication of all this data and upacking it will be any faster than reading it. I guess we'll find out!

**Reply by jessecarterMOOSE** [reply]
well I'll tell you what's NOT fast... calling `getFunctionByName` on every node :(

**Reply by jessecarterMOOSE** [reply]
If I instead create the map that I wanted (vector->Function) just after the broadcast while still in initialSetup by looping through the first map and creating a second that stores the Function pointer (from getFunctionByName),  any performance gains I got in the first place are lost :(

I'm doing something like:
```
for (auto x : _func_name_map)
  _func_map[x.first] = &getFunctionByName(x.second);
```
which is painfully slow! any better ideas?

**Reply by jessecarterMOOSE** [reply]
to be clear, I don't think the looping that's the problem, it's all the calls to getFunctionByName. For my test case, I've got about 15k functions to look up. Maybe I need to rethink this...

**Reply by lindsayad** [reply]
What are you using the `_func_map` for? Do you need the map? Or could you just loop over functions, regardless of name, and perform your operations?

Are you going to use all 15k functions on each process?

**Reply by jessecarterMOOSE** [reply]
I've got a bunch of field data from an external code that exists in two parts- piecewise linear function data (time-temperature for example, but there are others), and then which nodes need which functions. So the map is used in my AuxKernel to look up the Function for the current  node and evaluate at the current time.

**Reply by jessecarterMOOSE** [reply]
I have an Action that executes before this that takes that first set of data and creates a bunch of PiecewiseLinear Functions. This step is looking those up and evaluating them at the nodes

**Reply by jessecarterMOOSE** [reply]
Maybe I should combine the coordinate data with the time interpolation data and do it all in one go. Can I create PeacewiseLinear functions on the fly in my Auxkernel? Then I would have only floats in my file and could use the DelimitedFileReader instead of rolling my own that reads strings.

**Reply by GiudGiud** [reply]
Which one is the slow step? The creation of the Function everywhere or the evaluation of the 'right' function on every node?

Creating a Function on the fly on every node is not advisable. That will be slow. 
You want to read the data ahead of time. 

However if they are all piecewiselinear interpolations, maybe you dont need to create the functions. 
Use a map from Nodes to the interpolation data, and do the piecewise linear interpolation on the fly

**Reply by lindsayad** [reply]
`getFunction*` interfaces call through to the warehouse holding the function and sadly perform an O(N) search for a function matching the provided name. Consequently if you have thousands of functions and you are doing a lot of searches, I am not surprised that this is very slow.

**Reply by lindsayad** [reply]
It sounds like you are doing N searches, so this will be N^2 complexity

**Reply by lindsayad** [reply]
So I would say the quickest solution to solving the slowness, without considering whether to modify framework code, is to follow something like @GiudGiud's suggestion and not rely on `Functions`

**Reply by jessecarterMOOSE** [reply]
I'm thinking of creating LinearInterpolation objects right in InitialSetup() and putting them in a map with the coordinates as the key

**Reply by jessecarterMOOSE** [reply]
That's all my Functions are, nothing fancy

**Reply by GiudGiud** [reply]
Another option could be to do a 4D interpolation function:
x,y,z are classically the node/Qp position. And t instead of being time is the input of your LinearInterpolation

The idea would be to group all the linear interpolations in the same function, and they are naturally retrieved from accessing at the Qp position.

I m not sure we have that anywhere right now.

**Reply by jessecarterMOOSE** [reply]
In my case it actually is time that I'm interpolating.

**Reply by GiudGiud** [reply]
so you have a different 1D function for every point in space?
I reckon a general 4D function could do it, though you need the full cross terms to do it, not just the product of f(x,y,z)*g(t)

**Reply by jessecarterMOOSE** [reply]
> so you have a different 1D function for every point in space? 

That's the idea, yeah.

**Reply by jessecarterMOOSE** [reply]
@GiudGiud @lindsayad - Refactored my code so there are no `getFunction` calls (all data is in a single text file). Also since my data file contains only floats now I can use `DelimitedFileReader`. This works great! Guess the file read wasn't the bottleneck after all...

New problem: it is my understanding that `DelimitedFileReader` by default will run on all ranks unless a Communicator object is passed as an argument. When I try to do this, I get `Reinitializing Because of Geometric Search Objects...........................` for what seems like forever (never actually tried to wait it out yet).

**Reply by GiudGiud** [reply]
this is likely to be a bug. I would imagine some processes are stuck and never reach the geometric search

**Reply by lindsayad** [reply]
`DelmitedFileReader` will perform a read if 
1. the communicator is null or 
2. the communicator is non-null and you are on rank 0

As @GiudGiud said, it does sound like you may be attempting to perform communication at different spots in the code on different processes. Are you passing a `Communicator` to the `DelimitedFileReader` constructor?

**Reply by jessecarterMOOSE** [reply]
I'm using it like example here: https://mooseframework.inl.gov/source/utils/MooseUtils.html except I'm in initialSetup().

I'm assuming if I do something like:

```
MooseUtils::DelimitedFileReader reader(file_name, &_communicator);
reader.read();
std::vector<std::vector<Real>>data = reader.getData();
```

Then `data` would be available on all ranks.

**Reply by jessecarterMOOSE** [reply]
Which I think *does* work, but something is getting hung up later on, and it works fine in serial.

**Reply by jessecarterMOOSE** [reply, marked as answer]
Update: Got this to work if I explicitly do the file read on rank 0, get the data, then broadcast, i.e.:

```
std::vector<std::vector<double>> data;
if (processor_id == 0)
{
  MooseUtils::DelimitedFileReader reader(file_name);
  reader.read();
  data = reader.getData();
}
_communicator.broadcast(data);
```

**Reply by lindsayad** [reply]
That's really weird that you were getting a hang with
```
MooseUtils::DelimitedFileReader reader(file_name, &_communicator);
reader.read();
std::vector<std::vector<Real>>data = reader.getData();
```
but I'm glad that things are working for you now. If you have the time, I'd be interested in whether you could attach gdb to your hung processes and get backtraces to figure out where all the processes are, but no worries if you don't have the time. Also if you have `gstack` on your machine, that's even easier than attaching gdb.
